bridge/tracker/tracker.go

```diff
@@ -0,0 +1,307 @@
+package tracker
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/0xPolygon/polygon-edge/blockchain/storage"
+	"github.com/hashicorp/go-hclog"
+	"github.com/umbracle/go-web3"
+	"math/big"
+)
+
+type Config struct {
+	Confirmations uint64
+	RootchainWS   string
+	DBPath        string
+	ContractABIs  map[string][]string
+}
+
+//	cancellable context for tracker's listening mechanism
+type contextSubscription struct {
+	context context.Context
+	cancel  context.CancelFunc
+}
+
+//	done returns the contexts Done channel.
+func (c *contextSubscription) done() <-chan struct{} {
+	return c.context.Done()
+}
+
+//	Tracker represents an event listener that notifies
+//	for each event emitted by a smart contract on the rootchain.
+//	Events are collected from blocks at the desired depth (confirmations).
+type Tracker struct {
+	logger hclog.Logger
+
+	// required block confirmations
+	confirmations *big.Int
+
+	// event channel
+	eventCh chan []byte
+
+	// cancel subscription
+	ctxSubscription contextSubscription
+
+	//	rootchain subscription object
+	sub subscription
+
+	//	rootchain client
+	client *rootchainClient
+
+	//	db for last processed block number
+	db storage.Storage
+
+	//	events that tracker will listen for
+	contracts []*contractABI
+}
+
+//	NewEventTracker returns a new tracker object.
+func NewEventTracker(logger hclog.Logger, config *Config) (*Tracker, error) {
+	if config == nil {
```

Comment:
Actually nothing we can do if config is nil. You can return error or change `config` type to `Config` (value type) instead of using default value